xcm/src/v1/multilocation.rs

```diff
@@ -323,6 +323,60 @@ impl MultiLocation {
 		}
 		Ok(())
 	}
+
+	/// Mutate `self` so that it represents the same location from the point of view of `target`.
+	/// The context of `self` is provided as `ancestry`.
+	///
+	/// Does not modify `self` in case of overflow.
+	pub fn reanchor(&mut self, target: &MultiLocation, ancestry: &MultiLocation) -> Result<(), ()> {
+		// TODO: Optimize this.
+
+		// 1. Use our `ancestry` to figure out how the `target` would address us.
+		let inverted_target = ancestry.inverted(target)?;
+
+		// 2. Prepend `inverted_target` to `self` to get self's location from the perspective of
+		// `target`.
+		self.prepend_with(inverted_target).map_err(|_| ())?;
+
+		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
+		// strictly needed.
+		self.simplify(target.interior());
+
+		Ok(())
+	}
+
+	/// Treating `self` as a context, determine how it would be referenced by a `target` location.
+	pub fn inverted(&self, target: &MultiLocation) -> Result<MultiLocation, ()> {
+		use Junction::OnlyChild;
+		let mut ancestry = self.clone();
+		let mut junctions = Junctions::Here;
+		for _ in 0..target.parent_count() {
+			junctions = junctions
+				.pushed_with(ancestry.interior.take_last().unwrap_or(OnlyChild))
```

Comment:
I'm not sure about this one, it's taking the last junction from `ancestry` and pushing it to `junctions`?

So let's assume we're working with a smart contract and it's trying to see how a parachain will reference it -- we'll have `Parachain(1050)/PalletIndex(60)/GeneralIndex(1)` as `ancestry` and `../../../Parachain(2042)` as `target`.

This loop here will loop 3 times, and on the each iteration, it'll take the last junction from `ancestry` and append to `junctions`. When the loop finishes, we'd get `GeneralIndex(1)/PalletIndex(60)/Parachain(1050)` as a result for `junctions`, which doesn't sound right to me, aren't we supposed to keep the order of the junctions intact?

I think this is what's supposed to happen:
```suggestion
				.pushed_front_with(ancestry.interior.take_last().unwrap_or(OnlyChild))
```